### ISSUE TITLE
[Feature] Is there a way to pause/debug tests on fail?

### ISSUE THREAD

**Issue by rhemfelt** [microsoft/playwright#27635]
In TestCafe, there's a simple `--debug-on-fail` command line flag that automatically pauses the test if it fails, so you can troubleshoot and debug it. It is singularly one of the most used and useful command line flags for me in TestCafe, I run it with at least 95% of my test runs.  

It doesn't seem that Playwright has this pause on fail option, making troubleshooting test failures a big pain. Trying to pinpoint exactly where in the test code it failed, then manually adding an `await page.pause();` before that action or assertion, then rerunning each failed test, then manually performing the action that ostensibly triggered the initial failure.   

Is there any way (outside of VS Code) to automatically pause a test on failure?

**Comment by aalexgabi** [comment]
This would save so much time. I really don't understand why is not the default behavior with --debug. See a related discussion that died here https://github.com/microsoft/playwright/issues/10132

When a test fails currently you need to:
- find the fie that failed
- open the file
- understand the code and which of the steps was failing by matching the steps in the ui against the code which is super hard
- add await page.pause()
- rerunning test

This happens very often and every time a minute or so is wasted.

**Comment by guido-visser** [comment]
This feature would be very welcome! I tried to use the Break on (un)caught exceptions, but that triggers too much. I just want to break when an assertion fails, so I can debug whats going on and fix the test that way instead of using the trace viewer. It's very nice, but breaking in debug mode would be better :)

**Comment by MaksimStepanenko** [comment]
Fully agree. I'm migrating my tests from TestCafe and I was surprised with this Playwright Debug behaviour. I do expect that I can debug my test on failure when I click on Debug test button. Otherwise I don't see any differences between just run the test or debug it.

**Comment by lolmaus** [comment]
This feature request is over a year old now. 😔 

The problem it suggests to fix is so painful that I would rather consider it a bug report! 😩

**Comment by lolmaus** [comment]
CC @yury-s

**Comment by elyran** [comment]
+1. Much needed.

A way around, not perfect, but you can at least stop / debug where the error occurs (after you know there's a problem at certain line of code / block):
```
async function withDebugger(fn: () => Promise<any>) {
  try {
    return await fn();
  } catch (error) {
    debugger;
    throw error;
  }
}

// Usage, within the test:
    await withDebugger(async () => {
      await page.getByLabel('Label').fill('Test Field 1');
    });
```

This will stop the test where the problem is.

Remember to run with Node's `--inspect` flag, and a Chrome debugger in the background.

**Comment by guido-visser** [comment]
Up

**Comment by tschuehly** [comment]
Still would be nice in 2025!

EDIT:
This is possible using JUNIT 5 Testwatchers. Here's an example in Kotlin

```
class PlaywrightDebugExtension : TestWatcher {
    override fun testSuccessful(context: ExtensionContext) {
        val testInstance = context.testInstance.orElse(null)
        if (testInstance is PlaywrightBase) {
            closeBrowser(context, testInstance)
        }
    }

    override fun testFailed(context: ExtensionContext, cause: Throwable) {

        val testInstance = context.testInstance.orElse(null)
        if (testInstance is PlaywrightBase) {
            if (System.getenv("PIPELINE") != "true") {
                testInstance.log.error("Test failed: ${cause.message}")
                testInstance.log.info("Pausing browser for debugging...")
                testInstance.page.pause()
            } else {
                closeBrowser(context, testInstance)
            }
        }
    }

    private fun closeBrowser(
        context: ExtensionContext,
        testInstance: PlaywrightBase
    ) {
        val path = Paths.get("playwright/${context.displayName}.zip")
        testInstance.context.tracing().stop(
            StopOptions().setPath(path)
        )
        testInstance.context.close()
    }
}
```

**Comment by florianschepp** [comment]
Yeah would love this. Currently need to run with --debug which just fails followed by then going into the specific test to add an await page.pause() manually right before the failing expect(). Would be great if there is a --debug --pause-on-fail combo.

**Comment by jedwards1211** [comment]
If I had a choice between the big fancy Playwright UI and pause on failure, I'd take pause on failure any day 😂

So much investment in big complex features that I haven't found very helpful in debugging common test failures...

**Comment by mogelbrod** [comment]
+1 on this - being able to opt in to on test failure automatically either:
- Trigger `debugger` while having access to as much of the test scope as possible (`page` as a minimum?)
- Trigger `page.pause()` so that it's possible to eval code/locators from the attached debugger (vscode from example)

**Comment by Skn0tt** [comment]
In 1.58, try running with `PWPAUSE=1` and let us know what you think. We've not released it publically yet because it might change a little under the hood, and it might change shape completely, but we'd love to hear your feedback.

**Comment by abouabdallaoui-pass** [comment]
Hi, 
Just tried PWPAUSE=1, and it seems to pause tests even if they did not fail.
Launching all the tests in a suite with the flag will execute the first one, and then display the message `Paused at test end. Press Ctrl+C to end.` without having the ability to do anything.
Am I missing something?

**Comment by Skn0tt** [comment]
You're not missing something, that's how it's implemented right now. Our thinking is that this feature is only ever useful for debugging a single test anyways, and in that case you want it to pause on both success and failure.
You're using it with multiple tests so clearly we're missing something! Could you elaborate on your usecase?

**Comment by tschuehly** [comment]
@Skn0tt 
I'm using it before pushing my changes and I execute all tests and then it stops on failure and I can check what the issue is and quickly fix it

**Comment by abouabdallaoui-pass** [comment]
@Skn0tt something like OP described. Only pause (and show de debug window) if any test fail. In any other case, continue.

**Comment by SinaKarimi7** [comment]
This flag/feature is very essential for developers to find/debug the tests quickly. Is there any plan to have this asap?

**Comment by gross-metaventis** [comment]
I heavily rely on this, when debugging flaky tests with --repeat-each.

I run playwright on an Ubuntu environment and implemented this a s a fixture:
```
    PauseOnFailure: [
        async ({ page }, use, testInfo) => {
            await use();

            // only handle failures when the test did not pass _and_ its status is unexpected 
            if (testInfo.status == 'passed' || testInfo.expectedStatus == testInfo.status) {
                return;
            }

            await beep(); // process.stderr.write('\x07') 

            // I set env.TEST_IS_HEADED in playwright.config.ts when command line option "--headed" is present
            const runsHeadless = (testInfo.project.use.headless ?? true) && env.TEST_IS_HEADED !== 'true';
            if (process.env.PAUSE_HEADED_ON_FAILURE !== 'true' || !runsHeaded) {
                return;
            }

            // since the playwright inspector does not display the error, we need to dump the stack ourselves
            console.info('TEST PAUSED!');
            console.info(`expected test status: ${testInfo.expectedStatus}`);
            console.info(`received test status: ${testInfo.status}`);
            console.error(`
${testInfo.errors.length == 0
  ? `No error details available.`
  : `Test failed with error message(s):\n`
}
${testInfo.errors.map(e => {
  const msg = e.message ?? e.value ?? Object.entries(e);
  const stack = e.stack ?? '';
  if (typeof msg == 'string' && stack.startsWith(msg)) {
    return stack;
   } else {
    return msg + (e.stack ? '\n'+stack : '')
   }
}).join('\n\n')}
`);
            await page.pause();
        },
        { auto: true },
    ],
```

I added this to my playwright.config.ts:
```
import _ from 'lodash';

// usual config goes here

if (env.TEST_IS_HEADED === undefined) {
    env.TEST_IS_HEADED = process.argv.includes('--headed')
}

// The main issue with this is Webkit, because of a race condition(?) with 
// the page object - the browser just hangs on an empty page.
// Resolved by disabling webkits trace settings, if pause-on-failure applies:
const webkit = config.projects?.filter(p=>p.name === 'webkit');
if (webkit?.length == 1) {
    const runsHeaded = env.TEST_IS_HEADED === 'true' || config.use?.headless === false || webkit[0].use?.headless === false;
    if (process.env.PAUSE_HEADED_ON_FAILURE === 'true' && runsHeaded) {
        _.merge(webkit[0], { use: {
                trace: 'off',
                screenshot: 'off',
                video: 'off',
        } } );
    }
}
```
Code is "anonymized", so maybe a few errors got in. But you get the picture. It might not be pretty, but it does the trick!
